src/output_mqtt.c

```diff
@@ -0,0 +1,434 @@
+    }
+}
+
+static void mqtt_client_free(struct mg_mgr *mgr)
+{
+    free(mgr->user_data);
+    mgr->user_data = NULL;
+    mg_mgr_free(mgr);
+}
+
+/* Helper */
+
+/// clean the topic inplace to [-.A-Za-z0-9], esp. not whitespace, +, #, /, $
+static char *mqtt_sanitize_topic(char *topic)
+{
+    for (char *p = topic; *p; ++p)
+        if (*p != '-' && *p != '.' && (*p < 'A' || *p > 'Z') && (*p < 'a' || *p > 'z') && (*p < '0' || *p > '9'))
+            *p = '_';
+
+    return topic;
+}
+
+/* MQTT printer */
+
+typedef struct {
+    struct data_output output;
+    struct mg_mgr *mgr;
+    char topic[256];
+    char *devices;
+    char *events;
+    char *states;
+    //char *homie;
+    //char *hass;
+} data_output_mqtt_t;
+
+static void print_mqtt_array(data_output_t *output, data_array_t *array, char *format)
+{
+    data_output_mqtt_t *mqtt = (data_output_mqtt_t *)output;
+
+    char *orig = mqtt->topic + strlen(mqtt->topic); // save current topic
+
+    for (int c = 0; c < array->num_values; ++c) {
+        sprintf(orig, "/%d", c);
+        print_array_value(output, array, format, c);
+    }
+    *orig = '\0'; // restore topic
+}
+
+static char *append_topic(char *topic, data_t *data)
+{
+    if (data->type == DATA_STRING) {
+        *topic++ = '/';
+        strcpy(topic, data->value);
+        mqtt_sanitize_topic(topic);
+        topic += strlen(data->value);
+    }
+    else if (data->type == DATA_INT) {
+        *topic++ = '/';
+        topic += sprintf(topic, "%d", *(int *)data->value);
+    }
+    else {
+        fprintf(stderr, "Can't append data type %d to topic\n", data->type);
+    }
+
+    return topic;
+}
+
+// <prefix>/[<type>/][<model>/][<subtype>/][<channel>|<id>/]battery: "OK"|"LOW"
+static void print_mqtt_data(data_output_t *output, data_t *data, char *format)
+{
+    data_output_mqtt_t *mqtt = (data_output_mqtt_t *)output;
+
+    char *orig = mqtt->topic + strlen(mqtt->topic); // save current topic
+    char *end  = orig;
+
+    // collect well-known top level keys
+    data_t *data_type    = NULL;
+    data_t *data_brand   = NULL;
+    data_t *data_model   = NULL;
+    data_t *data_subtype = NULL;
+    data_t *data_channel = NULL;
+    data_t *data_id      = NULL;
+    for (data_t *d = data; d; d = d->next) {
+        if (!strcmp(d->key, "type"))
+            data_type = d;
+        else if (!strcmp(d->key, "brand"))
+            data_brand = d;
+        else if (!strcmp(d->key, "model"))
+            data_model = d;
+        else if (!strcmp(d->key, "subtype"))
+            data_subtype = d;
+        else if (!strcmp(d->key, "channel"))
+            data_channel = d;
+        else if (!strcmp(d->key, "id"))
+            data_id = d;
+    }
+
+    // top-level only
+    if (!*mqtt->topic) {
+        // "events" topic
+        if (mqtt->events && data_model) {
+            char message[1024];
+            data_print_jsons(data, message, sizeof(message));
```

Comment:
This buffer size needs a check. 1024 bytes might not be enough for really large messages.